libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,567 @@
+	vector<optional<StackSlot>> idealLayout(_post.size(), nullopt);
+	for (auto const& [slot, idealPosition]: ranges::zip_view(_post, layout))
+		if (PreviousSlot* previousSlot = std::get_if<PreviousSlot>(&idealPosition))
+			idealLayout.at(previousSlot->slot) = slot;
+
+	while (!idealLayout.empty() && !idealLayout.back())
+		idealLayout.pop_back();
+
+	return idealLayout | ranges::views::transform([](optional<StackSlot> s) {
+		yulAssert(s, "");
+		return *s;
+	}) | ranges::to<Stack>;
+}
+}
+
+Stack StackLayoutGenerator::propagateStackThroughOperation(Stack _exitStack, CFG::Operation const& _operation)
+{
+	Stack& stack = _exitStack;
+
+	vector<set<unsigned>> targetPositions(_operation.output.size(), set<unsigned>{});
+	size_t numToKeep = 0;
+	for (size_t idx: ranges::views::iota(0u, targetPositions.size()))
+		for (unsigned offset: findAllOffsets(stack, _operation.output.at(idx)))
+		{
+			targetPositions[idx].emplace(offset);
+			++numToKeep;
+		}
+
+	auto layout = ranges::views::iota(0u, stack.size() - numToKeep) |
+		ranges::views::transform([](size_t _index) { return PreviousSlot{_index}; }) |
+		ranges::to<vector<variant<PreviousSlot, set<unsigned>>>>;
+	// The call produces values with known target positions.
+	layout += targetPositions;
+
+	stack = createIdealLayout(stack, layout);
+
+	if (auto const* assignment = get_if<CFG::Assignment>(&_operation.operation))
+		for (auto& stackSlot: stack)
+			if (auto const* varSlot = get_if<VariableSlot>(&stackSlot))
+				if (util::findOffset(assignment->variables, *varSlot))
+					stackSlot = JunkSlot{};
+
+	for (StackSlot const& input: _operation.input)
+		stack.emplace_back(input);
+
+	m_layout.operationEntryLayout[&_operation] = stack;
+
+	// TODO: We will potentially accumulate a lot of return labels here.
+	// Removing them naively has huge implications on both code size and runtime gas cost (both positive and negative):
+	//   cxx20::erase_if(*m_stack, [](StackSlot const& _slot) { return holds_alternative<FunctionCallReturnLabelSlot>(_slot); });
+	// Consider removing them properly while accounting for the induced backwards stack shuffling.
+
+	// Remove anything from the stack top that can be freely generated or dupped from deeper on the stack.
+	while (!stack.empty() && (
+		canBeFreelyGenerated(stack.back()) ||
+		util::findOffset(stack | ranges::views::drop_last(1), stack.back())
+	))
+		stack.pop_back();
+
+	// TODO: suboptimal. Should account for induced stack shuffling.
+	// TODO: consider if we want this kind of compression at all, resp. whether stack.size() > 12 is a good condition.
+	if (stack.size() > 12)
+		stack = stack | ranges::views::enumerate | ranges::views::filter(util::mapTuple([&](size_t _index, StackSlot const& _slot) {
+			// Filter out slots that can be freely generated or are already present on the stack.
+			return !canBeFreelyGenerated(_slot) && !util::findOffset(stack | ranges::views::take(_index), _slot);
+		})) | ranges::views::values | ranges::to<Stack>;
```

Comment:
This is also important, but I think it should be easier to do properly after https://github.com/ethereum/solidity/pull/11493/files#r651807831

Reply:
This snippet of code is the main source of non-optimality on the one hand, on the other hand, just removing it will have the stack explode, since it will accumulate return labels and constants without check.

One option is to properly fix https://github.com/ethereum/solidity/pull/11493/files#r651807831 to be able to account for the stack shuffling implicit in removing slots here.

Another option is to drop this here entirely - but that would mean fixing more cases in ``fixStackTooDeep``, because doing nothing here will result in significantly larger stacks.

Reply:
Maybe an example:
We might be left with ``[0x42 0x21 0x07 a b]`` as stack here. Now assume 5 > 12 ;-), so we want to remove the constants in there.
Currently, they are just dropped, so we get:
``[a b]``.
Now suppose you want to get back to the layout you *actually* want, i.e. from ``[a b]`` to ``[0x42 0x21 0x07 a b]``. One way would be:
- ``[a b]``
- ``[a b 0x42]``
- ``[0x42 b a]``
- ``[0x42 b a 0x21]``
- ``[0x42 0x21 a b]``
- ``[0x42 0x21 a b 0x07]``
- ``[0x42 0x21 0x07 b a]``
- But now we have ``b a`` not ``a b`` in the end, so we need an additonal swap.

So in that example, we should not have reduced ``[0x42 0x21 0x07 b a]`` to ``[b a]`` but to ``[a b]`` instead to avoid the additional swap.
Note that the resulting penalty can be much higher than a single slot in more complex cases.

Conceptually, this can be done quite similarly to ``createIdealLayout`` above, but the concrete implementation won't cut it for these cases.
Similar we'd also want to be able to reduce ``[b b a]`` to ``[a b]`` or ``[b a]``, i.e. collapse duplicates.

Also as the comment says: the complete stack growing too large is probably not a good condition for doing this. Ideally we would just keep the stack growing, until it causes problems and compress only then.

Reply:
About literals, I think the same idea as before can be applied here, but this case is simpler.

Let's say that stack after is `Y`. If the stack before is `[X, 1, 2, ..., n]` were `1, ..., n` are literals. The idea is to permute only `1, ..., n` to the right place in `Y`. Now, apply the same permutation on `Y`,  which would transform `Y` to `[Z, 1, 2, ..., n]`. Now `X = Z` is what we want. For computing just permutations, this should do: https://gist.github.com/hrkrshnn/382a28656e56c33dfbb957cc255617f7. I know you mentioned that a simple recursive algorithm is also enough.

Reply:
The thing is: the literals don't have to be distinct and if they aren't dupping is better than pushing again... which leads back to the same situation as https://github.com/ethereum/solidity/pull/11493/files#r651807831.
I mean, if it turns out to be *exceedingly* problematic to do properly handle dups, we can just leave it, but my guess would be that it should be fine, it's just a bit more of a hassle :-).